wandb/sdk/launch/runner/kubernetes.py

```diff
@@ -0,0 +1,222 @@
+            time.sleep(1)
+            timeout -= 1
+            job_response = self.batch_api.read_namespaced_job_status(name=self.name, namespace=self.namespace)
+        
+        if timeout == 0 or job_response.status.conditions[0].type != "Suspended":
+            raise LaunchError("Failed to suspend job {}. Check Kubernetes dashboard for more info.".format(self.name))
+
+    def cancel(self) -> None:
+        self.suspend()
+        # todo: is that right? suspension will already delete all active pods but suspended jobs can be resumed/uncancelled - should we delete the job
+        
+
+class KubernetesRunner(AbstractRunner):
+    def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
+        kubernetes = get_module("kubernetes", "KubernetesRunner requires kubernetes to be installed")
+
+        resource_args = launch_project.resource_args.get("kubernetes")
+        if resource_args is None:
+            raise LaunchError("No Kubernetes resource args specified. Specify args via --resource-args with a JSON file or string under top-level key kubernetes")
+        
+        config_file = resource_args.get("config_file")   # kubeconfig, if None then loads default in ~/.kube
+        kubernetes.config.load_kube_config(config_file)
+
+        batch_api = kubernetes.client.BatchV1Api()
+        core_api = kubernetes.client.CoreV1Api()
+        api_client = kubernetes.client.ApiClient()
+
+        # allow users to specify template or entire spec
+        if resource_args.get("job_spec"):
+            job_dict = load_json_yaml_dict(resource_args['job_spec'])
+        else:
+            # begin constructing job sped
+            job_dict = {'apiVersion': 'batch/v1', 'kind': 'Job'}
+
+        # extract job spec component parts for convenience
+        job_metadata = job_dict.get("metadata", {})
+        job_spec = job_dict.get("spec", {})
+        pod_template = job_spec.get("template", {})
+        pod_metadata = pod_template.get("metadata", {})
+        pod_spec = pod_template.get("spec", {})
+        containers = pod_spec.get("containers", [{}])
+        job_status = job_dict.get('status', {})
+
+        # begin pulling resource arg overrides. all of these are optional
+
+        # allow top-level namespace override, otherwise take namespace specified at the job level, or default
+        namespace = resource_args.get("namespace", job_metadata.get("namespace", "default"))
+
+        # name precedence: resource args override > name in spec file > generated name
+        job_metadata['name'] = resource_args.get("job_name", job_metadata.get('name'))
+        if not job_metadata.get('name'):
+            job_metadata['generateName'] = "launch-"
+
+        if resource_args.get('cluster_name'):
+            job_metadata['clusterName'] = resource_args.get("cluster_name")
+        if resource_args.get('job_labels'):
+            job_metadata['labels'] = resource_args.get('job_labels')
```

Comment:
are job labels optional

Reply:
Yep, labels are optional.

Reply:
But we might want to always add a label that can tell us it's one of "our jobs"